<a href="https://colab.research.google.com/github/kaushik530411/DTN/blob/master/Lab_1_Exploring_Your_Data_with_Movie_Lens.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### Copyright 2019 Google LLC.

In [0]:
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Exploring Your Data with Movie Lens

Pandas is a powerful and popular Python library that is used to analyze, manipulate, and better understand table data. In this colab we'll explore the Pandas toolkit for operations such as filtering, grouping, merging, analysis.


## Overview

###Learning Objectives
* Use Colab to write and execute Python code integrating with Google Drive. 
* Introduce the `DataFrame` and `Series` data structures from the *Pandas* library.
* Analyze the data with descriptive statistics, and visualize the data with graphics


### Estimated Duration

We expect you to spend 3-5 hours completing this assignment.  We strongly encourage you to complete at least the first three exercises by mid-day on Friday so you can get help if needed.

### Grading

Exercises 1-7 are each worth 1 point.  Exercise 8 is worth 3 points.

## Colab




Colaboratory (Colab) is a cloud-based version of [Jupyter Notebook](https://jupyter.org/), where notebooks (which contain explanatory text and runnable code blocks), are stored on Google Drive. Notebooks are easy to share, comment,and edit just like a Google Doc, but allows you to run code blocks interactively.  If you prefer,A colab can be exported to github or to a Jupyter file.

### Cells
A notebook is a list of cells. Cells contain either explanatory text or executable code and its output. Click a cell to select it.

#### Code cells

Below is a **code cell**. Once the toolbar button indicates CONNECTED, click in the cell to select it and execute the contents in one of the following ways:

* Click the **Play icon** in the left gutter of the cell;
* Type **Cmd/Ctrl+Enter** to run the cell in place;
* Type **Shift+Enter** to run the cell and move focus to the next cell (adding one if none exists); or
* Type **Alt+Enter** to run the cell and insert a new code cell immediately below it.

Select the **Runtime** menu to see additional options for running some or all cells 


In [0]:
# Print Hello World
print("Hello World")

# Do some basic computation
x = 5
y = x + 2
print(y)

# We can define Python functions
def add_two(x):
  return x+2
print(add_two(20))

Hello World
7
22


#### Text cells

Now that we've seen an example of a code cell, this is a **text cell**.

To edit this text cell, **double-click** on it now. In edit mode, the left side of the cell contains markdown code. Notice there are extra asterisk characters. That's because Colab text cells use markdown syntax (described more below).

Notice the icons in the upper left corner for formatting the text (size, bold, italics, format as code, link, image, indent, bullet, horizontal rule). You can either use an icon (e.g. B for bold) or you can type in the markdown code yourself (double asterisks to make the text bold). See the [markdown
guide](/notebooks/markdown_guide.ipynb). Experiment by italicizing a word. To exit edit mode, click outside the edit area. 



### Integration with Drive

Colaboratory is integrated with Google Drive. Multiple people can share, comment, and collaborate on the same notebook.

* The **SHARE** button (top-right of the toolbar) allows you to share the notebook and control permissions set on it.

* **File->Save a Copy**  creates a copy of the notebook on Drive or Github.

* **File->Save** saves the File to Drive. **File->Save and checkpoint** explicitly saves the file and will show up in the version history as pinned.  Checkpoints are also saved periodically but do not show up as pinned.  So by using this feature, you can make it easy to find a particlar checkpoint.

* **File->Revision history** shows the notebook's revision history. 

* Multiple people can **collaboratively edit** a notebook at the same time. Like Google Docs, you can see collaborators both within the document (top right, left of the comments button) and within a cell (right of the cell). 

### Commenting on a cell
You can comment on a Colaboratory notebook like you would on a Google Document. Comments are attached to cells, and are displayed next to the cell they refer to. If you have **comment-only** permissions, you will see a comment button on the top right of the cell when you hover over it.

If you have edit or comment permissions you can comment on a cell in one of three ways: 

1. Select a cell and click the comment button in the toolbar above the top-right corner of the cell.
1. Right click a text cell and select **Add a comment** from the context menu.
3. Use the shortcut **Ctrl+Shift+M** to add a comment to the currently selected cell. 

You can resolve and reply to comments, and you can target comments to specific collaborators by typing *+[email address]* (e.g., `+user@domain.com`). Addressed collaborators will be emailed. 

The Comment button in the top-right corner of the page shows all comments attached to the notebook.

### Running Cells
You can run a single cell just by clicking on the "play" button by it.  By using the `RunTime` menu you can run all code cells, run all code cells before the current cell, run all code cells after the current cell, and other options.  Take a look at what is available.  

Note that any function or variable that has been defined, will remain defined even if you delete the cell that was used to define it. To start fresh you can use `Restart and run all...` in the `RunTime` menu. This will be done automatically if you aren't actively using your colab.

## Introduction to Pandas

Data typically comes in tables -- rows and columns. [*Pandas*](http://pandas.pydata.org/) is a Python library that is widely-used to manipulate table data. It makes it easy to handle all sorts of pre-processing we might want before feeding data to ML algorithms. 
Although a comprehensive introduction to the *Pandas* API would span many pages, the core concepts are fairly straightforward, and we'll present them below. For a more complete reference, the [*Pandas* docs site](http://pandas.pydata.org/pandas-docs/stable/index.html) contains extensive documentation and many tutorials.

We start with some standard imports:

In [0]:
# To run this code, hover your mouse over the [ ] and click the arrow. 
import pandas as pd
import numpy as np

Download a csv file to your computer with movie metadata by selecting [this link](https://storage.cloud.google.com/amli/public/movies_metadata.csv?folder=true&organizationId=433637338589). This data came from [The Movie Database org (TMDB)](https://www.themoviedb.org/?language=en-US) via [Kaggle](https://www.kaggle.com/rounakbanik/the-movies-dataset). The code below will upload the file into your colab runtime environment by having you choose a file.  You only need to download the data once but each time your runtime starts up you will need to choose it again.

__Note__:  The file is ~30MB in size, expect the download and upload to take a few minutes depending on your internet connection speed.

In [0]:
# Import the files package
from google.colab import files

# Get a list of the csv files that have been uploaded into your colab
# environment.
csv_uploaded = !ls *.csv

# If the metadata file is not already in the colab enviroment, load it
if ('public_movies_metadata.csv' not in csv_uploaded):
  uploaded = files.upload()

Saving public_movies_metadata.csv to public_movies_metadata.csv


#### Loading the movies database into a Pandas Dataframe Object

The movies available in this dataset correspond with the movies that are listed in the [MovieLens 
Latest Full Dataset](https://grouplens.org/datasets/movielens/latest/).

Let's load the data and look at the [head](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.head.html) (first few rows) of the table. Note you may need to scroll right to see all the columns.

In [0]:
# Create a pandas [dataframe -> DataFrame] object for the dataset. Ignore
# columns with no data.
movies_df = pd.read_csv('public_movies_metadata.csv').dropna(axis=1, how='all')

# Show the first few rows.
movies_df.head()

belongs_to_collection  ...  vote_count
0  {'id': 10194, 'name': 'Toy Story Collection', ...  ...      5415.0
1                                                NaN  ...      2413.0
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...  ...        92.0
3                                                NaN  ...        34.0
4  {'id': 96871, 'name': 'Father of the Bride Col...  ...       173.0

[5 rows x 23 columns]

A Pandas DataFrame has many [attributes and functions (also known as methods)](https://pandas.pydata.org/pandas-docs/stable/reference/frame.html#attributes-and-underlying-data). For example, you can use the [`shape`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shape.html) attribute to get the dimensionality (# rows, # cols).

In [0]:
movies_df.shape

(45453, 23)

Twenty-three features of 45,453 movies is going be a lot to look at... so let's start by checking the column (feature) descriptions:

In [0]:
movies_df.columns

Index(['belongs_to_collection', 'budget', 'genres', 'homepage', 'id',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count'],
      dtype='object')

Here's an explanation of each column:
- __belongs_to_collection__: The collection that a movie belongs to (if any).
- __budget__: The budget of the movie in dollars.
- __genres__:  List of genres associated with the movie.
- __homepage__: The Official Homepage of the movie.
- __id__: An arbitrary ID for the movie.
- __imdb_id__: The IMDB ID of the movie.
- __original_language__: The language in which the movie was filmed.
- __original_title__: The title of the movie in its original language.
- __overview__: A blurb of the movie.
- __popularity__: The Popularity Score assigned by TMDB.
- __poster_path__: The URL of the poster image (relative to http://image.tmdb.org/t/p/w185/).
- __production_companies__: A stringified list of production companies involved with the making of the movie.
- __production_countries__: A stringified list of countries where the movie was filmed or produced.
- __release_date__: Theatrical release date of the movie.
- __revenue__: World-wide revenue of the movie in dollars.
- __runtime__: Duration of the movie in minutes.
- __spoken_languages__: A stringified list of spoken languages in the film.
- __status__: Released, To Be Released, Announced, etc.
- __tagline__: The tagline of the movie.
- __title__: The official title of the movie.
- __video__: Indicates if there is a video present of the movie with TMDB.
- __vote_average__: The average rating of the movie on TMDB.
- __vote_count__: The number of votes by users, as counted by TMDB.


### Dataframe and Series
A [`Dataframe`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) holds two-dimensional tabular with labeled axes (rows and columns). [Geek for Geeks has a nice pictoral view](https://www.geeksforgeeks.org/python-pandas-dataframe/).   A [`Series`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html) is a column or row in a `Dataframe`. Arithmetic operations  can be performed on a row or column.

In [0]:
# Let's look at the Series corresponding to a column using the column
# labels we have provided.
movies_df['spoken_languages']

0                 [{'iso_639_1': 'en', 'name': 'English'}]
1        [{'iso_639_1': 'en', 'name': 'English'}, {'iso...
2                 [{'iso_639_1': 'en', 'name': 'English'}]
3                 [{'iso_639_1': 'en', 'name': 'English'}]
4                 [{'iso_639_1': 'en', 'name': 'English'}]
                               ...                        
45448               [{'iso_639_1': 'fa', 'name': 'فارسی'}]
45449                    [{'iso_639_1': 'tl', 'name': ''}]
45450             [{'iso_639_1': 'en', 'name': 'English'}]
45451                                                   []
45452             [{'iso_639_1': 'en', 'name': 'English'}]
Name: spoken_languages, Length: 45453, dtype: object

In [0]:
# We can also access an individual entry
print(movies_df['spoken_languages'][25])

[{'iso_639_1': 'en', 'name': 'English'}]


### Column Arithmetic

We can do arithmetic on columns as if they were numbers. This applies to using multiple columns too:

In [0]:
movies_df['revenue'] - movies_df['budget']

0        343554033.0
1        197797249.0
2                0.0
3         65452156.0
4         76578911.0
            ...     
45448            0.0
45449            0.0
45450            0.0
45451            0.0
45452            0.0
Length: 45453, dtype: float64

The code below creates a new column called `profit`.

In [0]:
movies_df['profit'] = movies_df['revenue'] - movies_df['budget']

movies_df.head()

belongs_to_collection  ...       profit
0  {'id': 10194, 'name': 'Toy Story Collection', ...  ...  343554033.0
1                                                NaN  ...  197797249.0
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...  ...          0.0
3                                                NaN  ...   65452156.0
4  {'id': 96871, 'name': 'Father of the Bride Col...  ...   76578911.0

[5 rows x 24 columns]


### Filtering the rows 

We often need to look at only a subset of the data that we are provided with. This is called **filtering**. Pandas provides many ways to filter our data. **One approach is to use Python’s index operator [ ] with a boolean test inside the brackets.**

For example, to consider films that lost money, create a variable called `money_loser_df` that contains all columns for the movies with revenue  less than their budget.

In [0]:
# Filter the original data by negative profit.
money_loser_df = movies_df[movies_df.profit < 0]

print(money_loser_df.shape)
money_loser_df.head()

(5113, 24)


belongs_to_collection    budget  ... vote_count      profit
6                    NaN  58000000  ...      141.0 -58000000.0
13                   NaN  44000000  ...       72.0 -30318235.0
14                   NaN  98000000  ...      137.0 -87982678.0
19                   NaN  60000000  ...      224.0 -24568887.0
22                   NaN  50000000  ...      394.0 -19696928.0

[5 rows x 24 columns]

That's more than 5000 movies that lost money! Clearly a risky business.

The expression `movies_df.profit < 0` was converted to an array of True and False values by [NumPy](hhttps://en.wikipedia.org/wiki/NumPy) (the numerical Python library imported above, which specializes in numerical operations on matrices). Pandas takes that array and uses it to decide which rows to include in the output and which rows to exclude.


One weird quirk of filtering with booleans is that when our test has multiple clauses (like: A and B), we don't use the standard Python syntax of `and`, `or` and `not`. Instead we must use the bitwise operators which are `&`, `|` and `!` respectively and put parentheses around each clause. So, for example, to select all movies that lost money and cost more than 1 million to make, we would use:

In [0]:
expensive_failures = movies_df[(movies_df.profit < 0) & (movies_df.budget > 1000000)]

expensive_failures.head()

belongs_to_collection    budget  ... vote_count      profit
6                    NaN  58000000  ...      141.0 -58000000.0
13                   NaN  44000000  ...       72.0 -30318235.0
14                   NaN  98000000  ...      137.0 -87982678.0
19                   NaN  60000000  ...      224.0 -24568887.0
22                   NaN  50000000  ...      394.0 -19696928.0

[5 rows x 24 columns]

### Filtering by columns and rows

Datasets often contain columns we don't care about. We can filter columns by using the *double* bracket notation and specifying column names:

*TIP: To remember this, generally remember that single brackets filter the rows while double brackets filter the columns.*

In [0]:
expensive_failures[['title', 'budget', 'revenue']]

title    budget     revenue
6                               Sabrina  58000000         0.0
13                                Nixon  44000000  13681765.0
14                     Cutthroat Island  98000000  10017322.0
19                          Money Train  60000000  35431113.0
22                            Assassins  50000000  30303072.0
...                                 ...       ...         ...
45385  Thick Lashes of Lauri Mäntyvaara   1254040         0.0
45389                   Corporate Event   2000000         0.0
45395                              Mars   2000000         0.0
45399                       Pro Lyuboff   2000000   1268793.0
45409                           Antidur   5000000   1413000.0

[3792 rows x 3 columns]

Recall that you can access a column of a `Dataframe` as a [`Pandas Series`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html).  Let's do this for the `vote_average` column of `money_loser_df`.



In [0]:
pd.Series(money_loser_df['vote_average'])

6        6.2
13       7.1
14       5.7
19       5.4
22       6.0
        ... 
45386    6.0
45389    3.5
45395    5.0
45399    4.0
45409    1.0
Name: vote_average, Length: 5113, dtype: float64

You can get the values within a Series (without the index) using `values`.

In [0]:
money_loser_df['vote_average'].values

array([6.2, 7.1, 5.7, ..., 5. , 4. , 1. ])

Instead of using the current row indices of 0, 1, 2, ....  suppose that we'd like the use the `title` of the movie as the index. We can do this by simply specifying the index we'd like to use.

In [0]:
# Define a series using the values of the vote_average column with an index of the 'title'
vote_lookup = pd.Series(money_loser_df['vote_average'].values, index=money_loser_df['title'])

# Now lookup the vote average of 'Dead Presidents'
vote_lookup['Dead Presidents']

6.6

There are other types of filters that we can use. For example, all string predicates are accessible via a column attribute called `str`. We can then use any Python string manipulation methods on them.  For example, we can use the `startswith` predicate to find all movies that start with a particular string or letter.

In [0]:
starts_with_p = vote_lookup[vote_lookup.index.str.startswith('P')]
print(starts_with_p)

title
Pushing Hands           6.1
Prêt-à-Porter           5.3
Pharaoh's  Army         6.3
Pete's Dragon           6.4
Paths of Glory          8.2
                       ... 
Pioneer Heroes          8.0
Perfect Opposites       4.8
Payback: Straight Up    6.8
Platon                  4.8
Pro Lyuboff             4.0
Length: 160, dtype: float64


If you'd like, you can sort these by the title.

In [0]:
print(starts_with_p.sort_index())

title
P2                        5.7
PCU                       6.0
Paa                       6.6
Padre nuestro             6.7
Palm Trees in the Snow    7.3
                         ... 
Purple Violets            5.7
Pushing Hands             6.1
Pushing Tin               5.7
Puzzlehead                5.1
Pyaar Ka Punchnama 2      7.7
Length: 160, dtype: float64


### Loading rating and user data from a zipfile
So far we have just loaded metadata about the movies that are used in the [MovieLens](https://grouplens.org/datasets/movielens/) dataset.

There are a variety of ways in which data can be provided. The Movie Lens user and rating data is most easily loaded from the web as a zip file and then the individual csv files are extracted.

In [0]:
# Download MovieLens data.
print("Downloading movielens data...")
from urllib.request import urlretrieve
import zipfile

# We upload the zip file with the data and then unzip it.
urlretrieve("http://files.grouplens.org/datasets/movielens/ml-100k.zip", "movielens.zip")
zip_ref = zipfile.ZipFile('movielens.zip', "r")
zip_ref.extractall()
print("Done. Dataset contains:")
print(zip_ref.read('ml-100k/u.info').decode('latin-1'))

Done. Dataset contains:
943 users
1682 items
100000 ratings



Next we can load the ratings csv file into a Pandas DataFrame.  We provide the name to user for each of the columns, and can indicate the separator used in the csv file.

In [0]:
ratings_cols = ['user_id', 'movie_id', 'rating', 'unix_timestamp']
ratings_df = pd.read_csv(
    'ml-100k/u.data', sep='\t', names=ratings_cols, encoding='latin-1')
ratings_df.head(10)

user_id  movie_id  rating  unix_timestamp
0      196       242       3       881250949
1      186       302       3       891717742
2       22       377       1       878887116
3      244        51       2       880606923
4      166       346       1       886397596
5      298       474       4       884182806
6      115       265       2       881171488
7      253       465       5       891628467
8      305       451       3       886324817
9        6        86       3       883603013

It would be more convenient to see the time when the movie was rated as readable time. You can find how to do this using a search such as "How to convert a unix timestamp to a readable time in python?"

A nice thing about Pandas is you can easily apply a function to all entries of a column as illustrated below where we add a human-readable time.

In [0]:
import time
ratings_df['rating_time'] = ratings_df['unix_timestamp'].apply(time.ctime)
# Let's look at the top rows of the Dataframe this created
ratings_df.head()

user_id  movie_id  rating  unix_timestamp               rating_time
0      196       242       3       881250949  Thu Dec  4 15:55:49 1997
1      186       302       3       891717742  Sat Apr  4 19:22:22 1998
2       22       377       1       878887116  Fri Nov  7 07:18:36 1997
3      244        51       2       880606923  Thu Nov 27 05:02:03 1997
4      166       346       1       886397596  Mon Feb  2 05:33:16 1998

Next we load data about the users that provided the ratings.

In [0]:
users_cols = ['user_id', 'age', 'sex', 'occupation', 'zip_code']
users_df = pd.read_csv(
    'ml-100k/u.user', sep='|', names=users_cols, encoding='latin-1')
users_df.head()

user_id  age sex  occupation zip_code
0        1   24   M  technician    85711
1        2   53   F       other    94043
2        3   23   M      writer    32067
3        4   24   M  technician    43537
4        5   33   F       other    15213

## Exercise 1: Basic Filtering

Explore the users table to answer the following questions


*   What is the average age of the user?
*   Generate and print a list of all occupations of the users in this data set.
*   Create a students_df table that only includes users who list their occupation as a student, and compute the average age of these users.

To help answer questions like these Pandas provides a rich set of [DataFrame functions](https://pandas.pydata.org/pandas-docs/stable/reference/frame.html).  For this exercise you will find the [Computations/descriptive stats](https://pandas.pydata.org/pandas-docs/stable/reference/frame.html#computations-descriptive-stats) functions to be useful.  If you aren't easily finding a function to meet your goal, an online search can often point you in the right direction.

### Student Solution

In [0]:
# Your code goes here

### Answer Key

**Solution**

In [0]:
print(users_df['age'].mean())

In [0]:
print(users_df['occupation'].unique())

In [0]:
students_df = users_df[users_df.occupation == 'student']
students_df['age'].mean()

## Exercise 2: More Advanced Filtering

Is this exercise, you are to compare the statistics (budget, popularity, revenue, runtime, vote_average, vote_count and profit) among all movies to those movies whose title include some word or substring of your choice (e.g. "money").  You can make use of Pandas [`describe`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html#pandas.DataFrame.describe) function to do this.

In a text box below your code, answer how many movies titles have the substring you selected and what you observe about the difference in the mean between those movies and the comparable values across all movies.

Since a few of the movies in this dataset do not have a title, you will need to filter out movies without a title. You may find 
```
movies_df[movies_df['title'].isnull() == False]
```
to be useful since it is the subset of rows that include a title.

While it may seem odd that a dataset about movies would include movies without a title, most large datasets are messy and will often have key fields missing.

### Student Solution

In [0]:
# Your code goes here

In this text box, write just a few sentences giving your observations about the the statistics ofthe movies that contain the word or substring of your choice as compared to all movies.

FILL IN YOUR ANSWER HERE.

### Answer Key

**Solution**

In [0]:
# I decided to find the average budget information for all the movies that
# include the word money as compared to those that do not.

# Need to drop Nan indices
movies_title_df = movies_df[movies_df['title'].isnull() == False]
movies_title_df.describe()

In [0]:
title_includes_money_df = movies_title_df[movies_title_df['title'].str.contains("Money")]
title_includes_money_df.describe()

Note: The goal here is to have the students look at the statistics in the table and make some comments about it.

There are 58 movies that contain the word "money".  Interestingly while the movies with "money" in the title were more popular and had a slightly higher vote_average, they were significantly less profitable primarily because that had a larger budget.

## Grouping

It is often useful to group data by a categorical value. For example, we might choose to group movies by language, country or origin, movie studio, etc.  Pandas provides the [`GroupBy`](https://pandas.pydata.org/pandas-docs/stable/reference/groupby.html) function to achieve this goal.
Let's try it out with language: Note that the result is not a `DataFrame` but a `DataFrameGroupBy` object. This object has one row for each language represented in `df` but isn't a `DataFrame` because Pandas doesn't know what to do with the other columns. We need to tell Pandas how we want the entries in the other columns to be combined. Even in the case where all entries are the same (like status in the illustration below) or there's a single entry being "grouped" (like nl below), Pandas awaits our instructions:

![group by figure](https://storage.googleapis.com/amli/public/group_by.png)

In the simple case where we want to combine all values in the same way, we can call one of the combining functions directly on the `DataFrameGroupBy` object. The most common combining functions are `sum`, `mean` and `count`:

In [0]:
original_language_averages = movies_df.groupby('original_language').mean()

original_language_averages.head(20)

There are a few things to notice here:
- non-numeric columns (like status or overview) were dropped from the output
- the column that we grouped by (__original_language__) is no longer a column of our output `DataFrame` but is its index instead
- some of these values are nonsensical: what does the average of an id mean?

You can now lookup information for a particular language:

In [0]:
original_language_averages.loc["fr"]

For more fine-grained analysis, you can specify individual combining functions using the `agg` method. Below we look at the total vote count and average vote rating by language.

In [0]:
movies_df.groupby('original_language').agg({'vote_count': 'sum', 'vote_average': 'mean'}).head(20)

### Exercise 3: Grouping and Aggregating

Let's find the total budget for the movies in our data for recent years that has this information (we don't have a lot of budget info for really old films). To do this, let's create a `DataFrame` with a row for each year from 1990 through 2017 (inclusive) and one column: the sum of all budgets in that year.

Some of the movies do not have a year, and thus will need to be dropped before we can filter by year.  You can use the [`isnull`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.isnull.html) function to do filter out these rows.

To process the __release_date__ column to extract the year, you can use the [Pandas Series string slice function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.slice.html) Also, for a string `s` you might find `s.astype(int)` to be useful to convert a string to an integer.  



#### Student Solution

In [0]:
# Your code goes here

## Merging

Frequently, data comes from different sources and has to be merged into a single data frame. For example, let's say that I have some notes about some of these movies that I want to merge:

In [0]:
movie_ratings_df = pd.merge(movies_df, ratings_df, left_on='id', right_on='movie_id')
movie_ratings_df.head()

It would be interesting to look at both the number of user ratings for each movie as well as the average user rating from the MovieLens data. In particular, we can create a new Dataframe first grouping rows by the `title`, and then aggregate these entries computing a count of the number of ratings and the mean rating for each movie.  Further, we can sort a Dataframe based on the values in any column.

In [0]:
movie_stats_df = movie_ratings_df.groupby('title', as_index=False).agg({'rating': ['count', 'mean']})
# sort by rating average
movie_stats_df.sort_values([('rating', 'count')], ascending=False).head()

We could instead sort by mean ratings.

In [0]:
movie_stats_df.sort_values([('rating', 'mean')], ascending=False).head()

### Exercise 4: Exploring Movie Rankings

It would be nice to find which movies are highest rated among those that received a reasonable number of ratings. Let's look at a subset of the movie_stats Dataframe that only includes movies with at least 100 ratings and find the highest and lowest rated 5 movies.  Does this change much if you instead look at movies with at least 20 ratings?

#### Student Solution

In [0]:
# Your code goes here

In this text box, write just a few sentences responding to the changes you see when looking at movies with at least 20 ratings.

FILL IN YOUR ANSWER HERE.

### Exercise 5: Are All Ranking Sources Consistent?

Observe that the TMDB dataset of movies metadata (`movies_df`) includes a `vote_average` and `vote_count`.  How consistent are these ratings with respect to the user ratings in the MovieLens dataset (`movie_ratings_df`)?  Why would they rank the movies in a different order?  Find a way to look to compare these two rating systems for the top 10 movies from each rating system.

You should not need to write more than about 10 lines of code to complete this exercise.

#### Student Solution

In [0]:
# Your code goes here

## Visualizing the Data: Tools & Techniques

There are many tools and techniques to perform initial exploration of a dataset.  In this lab, we use several of commonly available tools and through hands-on examples - you can learn by following along and completing additional practice sections on your own.  Reference materials are provided below.

  *   [NumPy](http://www.numpy.org/) and [Pandas](https://pandas.pydata.org/) are two libraries for data manipulation and analysis.
  *   [Matplotlib](https://matplotlib.org/) provide methods to create plots and visualizations.  While we won't discuss that here, for those interested [Seaborn](https://seaborn.pydata.org/index.html) and [Altair](https://altair-viz.github.io/) provide tools to create more sophisticated plots and visualizations.

In [0]:
# Imports for pyplot and seaborn
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns

### Generating histograms

We can now dig deeper and visualize some data distributions for various columns.

We start by plotting some some simple single variable distributions: user age distribution and movie rating distribution

In [0]:
plt.title("Histogram of ages of the raters")
users_df['age'].hist(bins=30)

In [0]:
plt.title("Distribution of movie ratings")
plt.ylabel('Number of user ratings of that value')
plt.xlabel('rating')
ratings_df['rating'].hist()


An interesting thing to look at is a histogram of the number of ratings across the movies as well as the mean ratings. This can be easily done.

In [0]:
movie_stats_df.hist(bins=20, figsize=(15, 5))

### Exercise 6: Number of Ratings per User

Create plots like done with `movie_stats` showing the number of ratings per movie and mean (average) rating over movies, to show the number of movies rated by user and the average rating of the user.

How do these histograms compare to those done over movies?  What differences do you see, and why do you think those are occuring?

#### Student Solution

In [0]:
# Your code goes here

In this text box, write just a few sentences as to the differences you see and why you think they are occuring.

FILL IN YOUR ANSWER HERE.

## Exercise 7: Visualizations for Movie Budgets Over Time

Let's return to Excercise 3. Looking at how the movie budget varied over time would be much easier to see as a plot.  Given a `Series` or a `DataFrame` you can call `plot()` method to obtain a plot of column values with respect to their index.  Try that out for visualizing total movie budget in a given year.

#### Student Solution

In [0]:
#your code goes here

## Exercise 8: Framing a Machine Learning Problem

In this exercise, you will use your knowledge of the Movies data to frame a machine learning task of your choice. Given what you now know about all the feature columns available, specify your task by answering the following questions:

* What are the inputs (x) and the outputs (y)?
* How will you evaluate your prediction function?
* What is a simple logical baseline (no ML allowed!) you can use?
* How will you split the data to avoid mis-interpretting performance?
* How might your learned system be used and what sorts of issues can you anticipate?

Then, implement your baseline and evaluate it on the full dataset (no need to split the data into train and test for this exercise).



### Student Solution

In this text box, write just a few sentences to answer each question.
* What are the inputs (x) and the outputs (y)?

FILL IN YOUR ANSWER HERE.

* How will you evaluate your prediction function?

FILL IN YOUR ANSWER HERE.

* What is a simple logical baseline (no ML allowed!) you can use?

FILL IN YOUR ANSWER HERE.

* How will you split the data to avoid mis-interpretting performance?

FILL IN YOUR ANSWER HERE.

* How might your learned system be used and what sorts of issues can you anticipate?

FILL IN YOUR ANSWER HERE.

In [0]:
# Your code goes here
# To receive full credit for implementing your baseline, you can give simple
# baseline.  No more than 15 lines of code should be given.